src/Preprocessor.zig

```diff
@@ -731,136 +652,256 @@ fn expandFunc(pp: *Preprocessor, source: *ExpandBuf, start_index: *usize, macro:
+                        idx += 1;
+                        continue;
+                    };
+                    defer {
+                        for (args.items) |item| {
+                            pp.comp.gpa.free(item);
+                        }
+                        args.deinit();
+                    }
+
+                    var args_count = @intCast(u32, args.items.len);
+                    // if the macro has zero arguments g() args_count is still 1
+                    if (args_count == 1 and func_macro.params.len == 0) args_count = 0;
+
+                    // Validate argument count.
+                    const extra = Diagnostics.Message.Extra{ .arguments = .{ .expected = @intCast(u32, func_macro.params.len), .actual = args_count } };
+                    if (func_macro.var_args and args_count < func_macro.params.len) {
+                        try pp.comp.diag.add(.{ .tag = .expected_at_least_arguments, .loc = buf.items[idx].loc, .extra = extra });
+                        idx += 1;
+                        continue;
+                    }
+                    if (!func_macro.var_args and args_count != func_macro.params.len) {
+                        try pp.comp.diag.add(.{ .tag = .expected_arguments, .loc = buf.items[idx].loc, .extra = extra });
+                        idx += 1;
+                        continue;
+                    }
+                    var expanded_args = MacroArguments.init(pp.comp.gpa);
+                    defer expanded_args.deinit();
+                    try expanded_args.ensureCapacity(args.items.len);
+                    for (args.items) |arg| {
+                        var expand_buf = ExpandBuf.init(pp.comp.gpa);
+                        try expand_buf.appendSlice(arg);
+
+                        try pp.expandMacroExhaustive(tokenizer, &expand_buf, 0, expand_buf.items.len, false);
+
+                        expanded_args.appendAssumeCapacity(expand_buf.toOwnedSlice());
+                    }
+
+                    var res = try pp.expandFuncMacro(&func_macro, &args, &expanded_args);
+                    defer res.deinit();
+                    for (expanded_args.items) |arg| {
+                        pp.comp.gpa.free(arg);
+                    }
+                    var expansion_loc = func_macro.loc;
+                    expansion_loc.next = buf.items[idx].loc.next;
+                    for (res.items) |*tok| {
+                        if (buf.items[idx].loc.next) |ln| {
+                            try pp.markExpandedFrom(tok, ln.*);
+                        }
+                        try pp.markExpandedFrom(tok, func_macro.loc);
+                    }
+
+                    try buf.replaceRange(idx, macro_scan_idx-idx+1, res.items);
+                    // TODO: moving_end_idx += res.items.len - (macro_scan_idx-idx+1)
+                    // doesn't work when the RHS is negative (unsigned!)
+                    moving_end_idx = moving_end_idx + res.items.len - (macro_scan_idx-idx+1);
```

Comment:
Any suggestion about this? It was moving_end_idx -= res.items.len - (macro_scan_idx-idx+1);
But the rhs can become negative, which is harmless but zig doesn't like it. It works right now, but I don't know if this is the zig way